packages/svelte-ux/src/lib/utils/icons.ts

```diff
@@ -0,0 +1,14 @@
+
+import type { ComponentProps } from "svelte";
+import type { default as Icon } from "$lib/components/Icon.svelte";
+
+export type IconInput = ComponentProps<Icon>['data'] | ComponentProps<Icon>;
+export type IconData = ComponentProps<Icon>['data'];
+
+export function asIconData(v: IconInput): IconData {
+  return isIconComponentProps(v) ? v.data : v;
+}
+
+function isIconComponentProps(v: any): v is ComponentProps<Icon> {
+	return typeof(v) === "object" && v && typeof(v['classes']) === "object" && "path" in v;
```

Comment:
There is a [isLiteralObject()](https://github.com/techniq/svelte-ux/blob/main/packages/svelte-ux/src/lib/utils/object.ts#L4C27-L6) util that could help refine this a little.

I'm also not sure `typeof(v['classes']) === "object" && "path" in v` is the correct check.  

I wonder if `asIconData()` should just be... 

```ts
export function asIconData(v: IconInput): IconData {
  return ('data' in v) ? v.data : v;
}
```

I might be overlooking something though.

Reply:
> There is a [isLiteralObject()](https://github.com/techniq/svelte-ux/blob/main/packages/svelte-ux/src/lib/utils/object.ts#L4C27-L6) util that could help refine this a little.

I agree, that's better to use.

> I'm also not sure `typeof(v['classes']) === "object" && "path" in v` is the correct check.

When I was writing this initially, I had thought that `ComponentProps<Icon>` required an object with all the same properties of `Icon`, but I can see now that it actually is like a `Partial` wrapper around that, so yeah, this definitely would *NOT* be the right check. Good catch.

Since `IconDefinition` has a more consistent structure & because it's the only other supported "object" type, we should just check to see if the given `v` object does NOT have certain properties that all `IconDefinition` objects are required to have and that `Icon` lacks. Therefore, a single property check can, by process of elimination, validate which type we're dealing with.

> I wonder if asIconData() should just be...

In theory, you'd be right (if it were just JS), but in order for TypeScript to rule out the possibility that the `IconInput` parameter is actually a `ComponentProps<Icon>`, you either have to exhaustively check all possible properties (far more work than is actually necessary) or write a function that forcibly coerces the type understanding via a type guard return value (`<param> is <type>` return type).

Reply:
fwiw, there is a [typeGuards](https://github.com/techniq/svelte-ux/blob/main/packages/svelte-ux/src/lib/types/typeGuards.ts) utils with `hasProperty()`, etc.  Might not be useful here